README.md

# flutter_blue_peripheral

A new Flutter plugin project.

## Getting Started

This project is a starting point for a Flutter
[plug-in package](https://flutter.dev/developing-packages/),
a specialized package that includes platform-specific implementation code for
Android and/or iOS.

For help getting started with Flutter development, view the
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.

# 平台
iOS android

# 使用
```
dependencies:
  frccblue:
    git:
      url: https://github.com/frcc00/frccblue
```
      
      
# 调用
```
Frccblue.init(didReceiveRead:(MethodCall call){
      print(call.arguments);
      return Uint8List.fromList([11,2,3,4,5,6,7,8,9,10,]);
    }, didReceiveWrite:(MethodCall call){
      print(call.arguments);
    },didSubscribeTo: (MethodCall call){
      print(call.arguments);
//      Frccblue.peripheralUpdateValue()
    },didUnsubscribeFrom: (MethodCall call){
      print(call.arguments);
    },peripheralManagerDidUpdateState: (MethodCall call){
      print(call.arguments);
    });

Frccblue.startPeripheral("00000000-0000-0000-0000-AAAAAAAAAAA1", "00000000-0000-0000-0000-AAAAAAAAAAA2").then((_){});
```

# more
## peripheralManagerDidUpdateState
iOS上传状态
```
switch peripheral.state {
        case .unknown:
            print("未知的")
            state = "unknown"
        case .resetting:
            print("重置中")
            state = "resetting"
        case .unsupported:
            print("不支持")
            state = "unsupported"
        case .unauthorized:
            print("未验证")
            state = "unauthorized"
        case .poweredOff:
            print("未启动")
            state = "poweredOff"
            self.peripheralManager?.stopAdvertising()
        case .poweredOn:
            print("可用")
            state = "poweredOn"
```
android上传状态
```
"unknown"
"poweredOff"
"poweredOn"
```


由于iOS没有设备连上和断开连接的回掉，android有；所以统一要求中心设备订阅Characteristic。
那么didSubscribeTo表示设备连上，didUnsubscribeFrom表示设备断开。android端didUnsubscribeFrom会被触发2次，在设备主动取消订阅的情况下。


CHANGELOG.md

## 0.1.0

* TODO: Describe initial release.


ios/Classes/SwiftFlutterBluePeripheralPlugin.swift

import Flutter
import UIKit
import CoreBluetooth

public class SwiftFlutterBluePeripheralPlugin: NSObject, FlutterPlugin, CBPeripheralManagerDelegate {

    var TAG = "[FlutterBluePeripheralPlugin] "

    var peripheralManager: CBPeripheralManager?
    var c: CBCentralManager?
    var channel: FlutterMethodChannel?

    let centralDic: NSMutableDictionary = [:]
    let characteristicDic: NSMutableDictionary = [:]

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name:"flutter_blue_peripheral", binaryMessenger: registrar.messenger())
        let instance = SwiftFlutterBluePeripheralPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    private var Service_UUID: String = "00000000-0000-0000-0000-AAAAAAAAAAA1"
    private var Characteristic_UUID: String = "00000000-0000-0000-0000-AAAAAAAAAAA2"

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "getPlatformVersion" {
            result("iOS " + UIDevice.current.systemVersion)
        }
        if call.method == "startPeripheral" {
            print(TAG + "startPeripheral")
            let param = call.arguments as! Dictionary<String,String>
            Service_UUID = param["serviceUuid"]!
            Characteristic_UUID = param["characteristicUuid"]!
            peripheralManager = CBPeripheralManager.init(delegate: self, queue: .main)
        }
        if call.method == "stopPeripheral" {
            print(TAG + "stopPeripheral")
            peripheralManager?.stopAdvertising()
        }
        if call.method == "peripheralUpdateValue" {
            let param = call.arguments as! NSDictionary
            let centralUuid = param["centralUuid"] as! NSString
            // let characteristicUuid = param["characteristicUuid"] as! NSString
            let data = param["data"] as! FlutterStandardTypedData
            // peripheralManager?.updateValue(
            //     data.data,
            //     for: (characteristicDic[characteristicUuid]) as! CBMutableCharacteristic,
            //     onSubscribedCentrals: [centralDic[centralUuid] as! CBCentral]
            // )
            peripheralManager?.updateValue(
                data.data,
                for: (characteristicDic[0]) as! CBMutableCharacteristic,
                onSubscribedCentrals: [centralDic[centralUuid] as! CBCentral]
            )
        }
    }

    public func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        var state = "unknown"
        switch peripheral.state {
        case .unknown:
            state = "unknown"
        case .resetting:
            state = "resetting"
        case .unsupported:
            state = "unsupported"
        case .unauthorized:
            state = "unauthorized"
        case .poweredOff:
            state = "poweredOff"
            self.peripheralManager?.stopAdvertising()
        case .poweredOn:
            state = "poweredOn"
            setupServiceAndCharacteristics()
            let deviceName = UIDevice.current.name;
            self.peripheralManager?.startAdvertising(
                [CBAdvertisementDataServiceUUIDsKey: [CBUUID.init(string: Service_UUID)],
                CBAdvertisementDataLocalNameKey: deviceName,
            ])
        @unknown default:
            state = "unknown"
        }
        print(TAG + "peripheralManagerDidUpdateState:" + state)
        channel?.invokeMethod("peripheralManagerDidUpdateState", arguments: state)
    }

    private func setupServiceAndCharacteristics() {
        let serviceID = CBUUID.init(string: Service_UUID)
        let service = CBMutableService.init(type: serviceID, primary: true)
        let characteristicID = CBUUID.init(string: Characteristic_UUID)
        let characteristic = CBMutableCharacteristic.init(
            type: characteristicID,
            properties: [.read, .write, .notify],
            value: nil,
            permissions: [.readable, .writeable]
        )
        service.characteristics = [characteristic]
        characteristicDic[0] = characteristic
        self.peripheralManager?.add(service)
    }

    public func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didSubscribeTo characteristic: CBCharacteristic) {
        print(TAG + "didSubscribeTo:" + central.identifier.uuidString)
        centralDic[central.identifier.uuidString] = central
        channel?.invokeMethod("didSubscribeTo", arguments: [
            "centralUuid": central.identifier.uuidString,
            "characteristicUuid": characteristic.uuid.uuidString,
        ])
    }

    public func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didUnsubscribeFrom characteristic: CBCharacteristic) {
        print(TAG + "didUnsubscribeFrom:" + central.identifier.uuidString)
        centralDic[central.identifier.uuidString] = nil
        channel?.invokeMethod("didUnsubscribeFrom", arguments: [
            "centralUuid": central.identifier.uuidString,
            "characteristicUuid": characteristic.uuid.uuidString,
        ])
    }

    public func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveRead request: CBATTRequest) {
        channel?.invokeMethod("didReceiveRead", arguments: [
            "centralUuid": request.central.identifier.uuidString,
            "characteristicUuid": request.characteristic.uuid.uuidString,
        ],
        result: { (data) in
            if let da = data as? FlutterStandardTypedData {
                request.value = da.data
            }
            peripheral.respond(to: request, withResult: .success)
        })
    }

    public func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
        for request in requests {
            if let data = request.value {
                channel?.invokeMethod("didReceiveWrite", arguments: [
                    "centralUuid": request.central.identifier.uuidString,
                    "characteristicUuid": request.characteristic.uuid.uuidString,
                    "data": FlutterStandardTypedData.init(bytes: data),
                ])
            }
            request.value = nil
            peripheral.respond(to: request, withResult: .success)
        }
    }
}
